### ISSUE TITLE
Genetic data retrieval pattern

### ISSUE THREAD

**Issue by BoPeng** [vatlab/VAT#8]
Before we design an efficient data storage model, we need to analyze the retrieval patterns of genetic data, namely, what data are needed at what time and in what pattern. For example, 

1. When variant annotations are needed? Sample preparation (filtering) and/or analysis?
2. What data are needed for association analysis? 0/1/2? phase?
3. Is order of variants significant?
4. Are genotypes usually obtained in groups (of variants, of samples)?

A schema of how data are retrieved would help us design a storage model.

**Comment by gaow** [comment]
Just returned from a conference ...

Typically, genotype data are retrieved not just for association analysis. For instance:

1. Sample level genotype data summary: `vtools show genotypes` gives just genotype counts. We also use `vtools phenotype ... --from_stat` for more detailed counts. 
2. Sample level genotype QC: This is just operations above conditional on genotype annotations, for example we count only genotypes with certain depth of coverage or quality scores
3. Variant level genotype data / QC: all above but per variant not per sample, e.g. `vtools update variant --from_stat`. This is particularly relevant when we want to compute MAF conditional on genotype qualities. 
4. Genotype annotations summaries such as `vtools output variant "max(DP)" "min(DP)" "avg(DP)" "stdev(DP)" "lower_quartile(DP)" "upper_quartile(DP)"`
5. Sample level ti/tv ratio
6. Other genotype related analysis before association testing: 
   * global ancestry analysis, requiring only a subset of variants but they will have to be selected from the entire genome!
   * local ancestry analysis: lacking from VAT for now. But also need same data-set (or denser, but not all) compared to global ancestry

In sum, all that we currently have with `vtools update --from_stat` and `vtools phentoype --from_stats` will be heavily used, and should all be refactored and ported over to VAT 3.0 nicely.

Additionally limitations of storage of genotype in VAT 2.0:

1. Does not include phase -- this will be relevant for haplotype based analysis though we do not have any for now. We have proposed though.
2. Does not include imputed genotype (dosage). This is starting to hurt as currently with the growth of WGS data as reference panels, imputed genotype is more often used.

For variants storage related issues:

1. Annotation can be used at any time. For example some applications might want to start off with a subset of variants e.g. exons only; other applications may want to keep everything along the way then treat different types of variants differently in QC, analysis etc, e.g. indels/snps
2. I'm not sure if order of variant is significant. But at least it will be odd to see exported variants having wrong ordering. Also I can imaging perhaps for, say regulatory variants the closer to transcription start sites (TSS) the larger the regulartory effect size on average and that might be used as some annotations. It is safer to try to maintain the ordering.
3. "Are genotypes usually obtained in groups (of variants, of samples)?" Yes -- and I'm not getting the question: what's the alternative, if not obtained in groups of variants or samples?

**Comment by gaow** [comment]
I'm not sure if we want to support many types of data providers. At the variant level, VCF/BCF does great job and if we want to acquire blocks of data we should use *CF indexing. For annotations I find most people I know just run the ANNOVAR pipeline. We need to check how much overlap there is between ANNOVAR and our annotation and decide if we should contribute to ANNOVAR of what we  do not have and embed ANNOVAR better (for example leverage ANNOVAR to build our DB), or trying to keep maintaining our own annotations. 

At this point I think instead of re-inventing the wheel it seems more tempting to put things together under the VAT interface, but more hassle free. From a user's prospective I'd definitely trust well-established tools made specific for annotation, meta analysis, family based analysis, even association testing (`rvtests` from Goncalo's lab) which have been applied to large projects. We have been working a lot on the scope of VAT as a one stop shop for association (2011 - 2014), on making it a pipeline tool (2013 - 2016) and on creating a unified data infrastructure (via `sqlite` though not salable). We should focus on certain aspects of these and utilize/co-evolve with existing analysis tools if we do not want to stretch ourselves too thin.

**Comment by BoPeng** [comment]
> Sample level genotype data summary: vtools show genotypes gives just genotype counts. We also use vtools phenotype ... --from_stat for more detailed counts.

Phenotypes are relatively small even when we talk about tens of thousands of samples and hundreds of phenotypes. So I am at all concerned with how they are stored.

> Sample level genotype QC: This is just operations above conditional on genotype annotations, for example we count only genotypes with certain depth of coverage or quality scores

I suppose this can be done before association analysis. More specifically, we can ask the data provider to send genotypes with certain depth of coverage or quality scores. Now, because the data provider can cache subsets of genotypes with certain criteria, relatively slow retrieval of genotype annotation should be tolerable. 

> Variant level genotype data / QC: all above but per variant not per sample, e.g. vtools update variant --from_stat. This is particularly relevant when we want to compute MAF conditional on genotype qualities.
Genotype annotations summaries such as vtools output variant "max(DP)" "min(DP)" "avg(DP)" "stdev(DP)" "lower_quartile(DP)" "upper_quartile(DP)"

This is more troublesome because we first need to calculate `max(DP)` etc from genotype data and then filter variants using certain cutoff values. This might be the place where the existing variant tables and related algorithm will remain.

> global ancestry analysis, requiring only a subset of variants but they will have to be selected from the entire genome!

There should not be any problem if we select thousands of variants.

> local ancestry analysis: lacking from VAT for now. But also need same data-set (or denser, but not all) compared to global ancestry

> In sum, all that we currently have with vtools update --from_stat and vtools phentoype --from_stats will be heavily used, and should all be refactored and ported over to VAT 3.0 nicely.

I agree.

> Additionally limitations of storage of genotype in VAT 2.0:

> Does not include phase -- this will be relevant for haplotype based analysis though we do not have any for now. We have proposed though.

Agreed.

> Does not include imputed genotype (dosage). This is starting to hurt as currently with the growth of WGS data as reference panels, imputed genotype is more often used.

I believe these can be done separately as the resulting dataset could be used as a separate data source.

> For variants storage related issues:

> Annotation can be used at any time. For example some applications might want to start off with a subset of variants e.g. exons only; other applications may want to keep everything along the way then treat different types of variants differently in QC, analysis etc, e.g. indels/snps

With the set up unified variant IDs, we can define numerous subset of variants based on different annotations. The subset of variants will have their IDs and be cached for better performance.

> I'm not sure if order of variant is significant. But at least it will be odd to see exported variants having wrong ordering. Also I can imaging perhaps for, say regulatory variants the closer to transcription start sites (TSS) the larger the regulartory effect size on average and that might be used as some annotations. It is safer to try to maintain the ordering.

I suppose most operations could be performed without order and we can use variant annotation (chr, pos) to order them if needed.

> "Are genotypes usually obtained in groups (of variants, of samples)?" Yes -- and I'm not getting the question: what's the alternative, if not obtained in groups of variants or samples?

Basically I would like to define subsets of variants so that they (e.g. variants belong to the same gene) could be returned quickly.

**Comment by BoPeng** [comment]
> We should focus on certain aspects of these and utilize/co-evolve with existing analysis tools if we do not want to stretch ourselves too thin.

I partially agree with this, but the main goal of the redesign is to scale VAT to WGS data and whether or not we should rely on third-party tools depends on whether or not they work with our new data processing model.

**Comment by gaow** [comment]
I just want to clarify that `vtools phenotype ... --from_stats` is operation over genotype data not phenotype data. Unlike `vtools update` it operates at individual level. It used to be one major operation where VAT 2.0 has speed issues. Recall the first problem we realized with VAT 2.0 in genotype access was the slowness in random access (thus the use of SSD) with `sqlite`.
